### PR TITLE
node/state: graceful shutdown in the consensus state

### PR DESCRIPTION
**Pull request by JayT106**
The current consensus reactor will stop mechanism has a chance to cause the data missing with the indexer services. Therefore we introduce a graceful stop if the consensus module is trying to commit the block. The OnStop process will wait for the block commit to finish and then continue the shutdown process.

Also, emit the index services-related events when the node replays the final block.

Closes:#5002
Might solves:#5851

### CODE REVIEW

File: node/node.go [hunk 1]
```diff
@@ -1404,7 +1404,6 @@ func (n *Node) OnStart() error {
 
 // OnStop stops the Node. It implements service.Service.
 func (n *Node) OnStop() {
-	n.BaseService.OnStop()
```
Comment (cmwaters) [inline node/node.go:1407]: What was your reasoning for removing this?

File: state/execution.go [hunk 2]
```diff
@@ -532,18 +532,41 @@ func fireEvents(
...
+			if err := be.eventBus.PublishEventTx(types.EventDataTx{TxResult: abci.TxResult{
+				Height: block.Height,
+				Index:  uint32(i),
+				Tx:     tx,
+				Result: *(abciResponses.DeliverTxs[i]),
+			}}); err != nil {
+				logger.Error("failed publishing event TX", "err", err)
+			}
+		}
+	}
```
Comment (cmwaters) [inline state/execution.go:568]: Is it possible to use the `fireEvents` function instead for consistency

File: consensus/reactor.go [hunk 3]
```diff
@@ -196,6 +196,18 @@ func (r *Reactor) OnStart() error {
...
+	var c = 0
+	for (r.state.Step == cstypes.RoundStepCommit) && (c < 3) {
+		time.Sleep(time.Second)
+		c++
+	}
+
+	if c == 3 {
+		r.Logger.Error("the 3 secs block commit waiting timeout! The block commit might be failed!")
+	}
+
```
Comment (cmwaters) [inline consensus/reactor.go:210]: The idea behind this is that **only** if we are in the `CommitStep` do we wait for that step to finish, not any other step right? 

Could we not add a listener for a `RoundStepNewHeight` event and wait for that to fire before continuing

File: node/node.go (hunk 1 above)
Comment (JayT106) [reply node/node.go:1407]: The BaseService.OnStop() looks like doing nothing. 
https://github.com/tendermint/tendermint/blob/05bc2f344a20bd25beb4b6f9e6f41d2295d99492/libs/service/service.go#L180

And the `func (n *Node) OnStop()` be called by https://github.com/tendermint/tendermint/blob/05bc2f344a20bd25beb4b6f9e6f41d2295d99492/libs/service/service.go#L159

File: state/execution.go (hunk 2 above)
Comment (JayT106) [reply state/execution.go:568]: Sure.

File: node/node.go (hunk 1 above)
Comment (cmwaters) [reply node/node.go:1407]: Yeah it seems like you're right. I'm unsure why it was there in the first place then. I notice that `BaseService.OnStop()` get's called in other places as well.

File: consensus/reactor.go (hunk 3 above)
Comment (JayT106) [reply consensus/reactor.go:210]: Right, when the consensus is entering the commit stage, the Step will be updated
https://github.com/tendermint/tendermint/blob/05bc2f344a20bd25beb4b6f9e6f41d2295d99492/consensus/state.go#L1487
and if commit block success, the Step will be updated again
https://github.com/tendermint/tendermint/blob/05bc2f344a20bd25beb4b6f9e6f41d2295d99492/consensus/state.go#L1676

File: node/node.go (hunk 1 above)
Comment (JayT106) [reply node/node.go:1407]: It looks weird when I was looking at the OnStop call. So I thought maybe this line has been added accidentally.

File: consensus/reactor.go (hunk 3 above)
Comment (JayT106) [reply consensus/reactor.go:210]: Thanks for the suggestion, I will check it.

Comment (alexanderbez) [reply consensus/reactor.go:210]: Yeah would prefer to see if we can wait for an event or something to fire first.

File: consensus/reactor.go [hunk 4]
```diff
@@ -196,6 +200,19 @@ func (r *Reactor) OnStart() error {
...
+	// If the node is committing the new block, wait until it finished!
+
+	if r.state.GetRoundState().Step == cstypes.RoundStepCommit {
+		select {
+		case <-r.newHeightCh:
+		case <-time.After(3 * time.Second):
+			r.Logger.Error("the 3 secs block commit waiting timeout! The block commit might be failed!")
+		}
+	}
```
Comment (cmwaters) [inline consensus/reactor.go:212]: I don't think you need to create a new event. We're just wanting to wait till `RoundStepCommit` is done, so we should be able to use `EventNewRoundStep` to monitor if that happens or not

Comment (cmwaters) [reply consensus/reactor.go:212]: Also I'm wondering if it's better to do this in the `OnStop` func within `State` rather than the reactor

Comment (JayT106) [reply consensus/reactor.go:212]: Right, I will move it into the state.OnStop(). Thanks for the suggestion.

File: consensus/state.go [hunk 5]
```diff
@@ -411,6 +415,25 @@ func (cs *State) loadWalFile() error {
 
 // OnStop implements service.Service.
 func (cs *State) OnStop() {
+
+	// If the node is committing the new block, wait until it finished!
```
Comment (cmwaters) [inline consensus/state.go:419]:
```suggestion
	// If the node is committing a new block, wait until it is finished!
```

File: consensus/state.go [hunk 6]
```diff
@@ -411,6 +415,25 @@ func (cs *State) loadWalFile() error {
...
+	if cs.GetRoundState().Step == cstypes.RoundStepCommit {
+	wait:
+		for {
```
Comment (cmwaters) [inline consensus/state.go:422]: Is a for loop necessary if you are breaking out of it in every case. AFAICS this only loops once

File: consensus/state.go [hunk 7]
```diff
@@ -411,6 +415,25 @@ func (cs *State) loadWalFile() error {
...
+			select {
+			case rs := <-cs.onStopCh:
+				if rs != nil && rs.Step == cstypes.RoundStepNewHeight && rs.Round == 0 {
+					break wait
+				}
+			case <-time.After(time.Duration(cs.config.TimeoutCommit.Seconds())):
```
Comment (cmwaters) [inline consensus/state.go:428]: Is it possible to just do it like this
```suggestion
			case <-time.After(cs.config.TimeoutCommit):
```

File: state/execution.go [hunk 8]
```diff
@@ -551,18 +551,36 @@ func fireEvents(
 // ExecCommitBlock executes and commits a block on the proxyApp without validating or mutating the state.
 // It returns the application root hash (result of abci.Commit).
 func ExecCommitBlock(
+	be *BlockExecutor,
 	appConnConsensus proxy.AppConnConsensus,
 	block *types.Block,
 	logger log.Logger,
 	store Store,
 	initialHeight int64,
+	s State,
```
Comment (cmwaters) [inline state/execution.go:560]: Where is this used?

File: consensus/state.go [hunk 9]
```diff
@@ -411,6 +415,25 @@ func (cs *State) loadWalFile() error {
...
+				cs.Logger.Error(fmt.Sprintf("the %v secs block commit timeout!", cs.config.TimeoutCommit.Seconds()))
```
Comment (cmwaters) [inline consensus/state.go:429]:
```suggestion
				cs.Logger.Error("OnStop: timeout waiting for commit to finish", "time", cs.config.TimeoutCommit)
```

File: state/execution.go [hunk 10]
```diff
@@ -551,18 +551,36 @@ func fireEvents(
...
 ) ([]byte, error) {
-	_, err := execBlockOnProxyApp(logger, appConnConsensus, block, store, initialHeight)
+	abciResponses, err := execBlockOnProxyApp(logger, appConnConsensus, block, store, initialHeight)
 	if err != nil {
 		logger.Error("failed executing block on proxy app", "height", block.Height, "err", err)
 		return nil, err
 	}
 
+	if be != nil {
```
Comment (cmwaters) [inline state/execution.go:568]: Can we add a comment about this

File: consensus/replay.go [hunk 11]
```diff
@@ -469,9 +469,23 @@ func (h *Handshaker) replayBlocks(
 			assertAppHashEqualsOneFromBlock(appHash, block)
 		}
 
-		appHash, err = sm.ExecCommitBlock(proxyApp.Consensus(), block, h.logger, h.stateStore, h.genDoc.InitialHeight)
-		if err != nil {
-			return nil, err
+		if i == finalBlock {
+			// We emit events for the index services at the final block due to the sync issue when
+			// the node shutdown during the block committing status.
```
Comment (cmwaters) [inline consensus/replay.go:474]: Do you think it's possible for the block to emit the same events twice?

**Review by cmwaters — CHANGES_REQUESTED** [review]
I have requested a few changes but other than that, this looks like its almost done

File: state/execution.go (hunk 8 above)
Comment (JayT106) [reply state/execution.go:560]: at line 570, the 'ConsensusParams.Validator'

Comment (cmwaters) [reply state/execution.go:560]: Right, I couldn't see it

File: state/execution.go (hunk 10 above)
Comment (JayT106) [reply state/execution.go:568]: Sure!

File: consensus/state.go (hunk 6 above)
Comment (JayT106) [reply consensus/state.go:422]: You are right, I was thinking there are other events that will fall into the first case but didn't meet the condition.

File: consensus/replay.go (hunk 11 above)
Comment (JayT106) [reply consensus/replay.go:474]: Yes, if `mutateState` flag is true. I will revise it. Do you think we need fireEvent if the flag is false? If no need I can rollback the change in this part

**Review by cmwaters — APPROVED** [review]
Awesome stuff thanks 🎉

File: consensus/replay.go (hunk 11 above)
Comment (cmwaters) [reply consensus/replay.go:474]: I think its fine to leave this here if `mutateState` is false

Comment (JayT106) [reply consensus/replay.go:474]: Cool. Thanks!

**Pull request merged**